playbooks/fundamentals/template_development_guide.adoc

```diff
@@ -0,0 +1,898 @@
+    version: 1.4.7
+  creationTimestamp: null
+  name: openjdk18-web-basic-s2i
+----
++
+Once copied into a different, special purpose template, this metadata no longer makes much sense. Its likely best to remove it, or update relevant fields if you are planning to load the template into the Web Console.
+- **Avoid editing existing templates in place; always make a copy**
++
+This advice is primarily for those wanting to update the out of the box templates that ship with OpenShift. The canned set of templates typically gets rolled out any time a cluster is upgraded, which will override any edits made to the templates. Its best to export a template and rename it to something that can be easily differentiated, like `myorg-openjdk-basic`.
+
+== Templates & Infrastructure as Code (IaC)
+
+Templates give us a simple, yet powerful vehicle upon which we can build sophisticated and consistent automation of nearly everything we do with OpenShift. In this section we propose some essential components of an automated workflow around OpenShift, and introduce an Ansible framework that can be used to implement them.
+
+=== Use oc apply for repeatable process
+
+We already discovered the value of `oc process | oc apply` during our template exploration at the beginning of this document. In general, `oc apply` carries a lot of value over some of the other alternatives such as `oc create`, `oc replace`, or `oc new-app`. Here are some things you should know about apply.
+
+- _apply_ only activates a trigger if a change is detected. This prevents `builds` and `deployments` from kicking off unnecessarily.
+- _apply_ will save a copy of the previous version of the object that was applied in the annotation `kubectl.kubernetes.io/last-applied-configuration`
+- _apply_ is getting heavy investment in the Kubernetes community
+
+This presentation from KubeCon 2017 provides more interesting deep dives into using `oc apply`.
+
+video::CW3ZuQy_YZw[youtube]
+
+=== Source Control for Templates
+
+Your templates should be version controlled. This cannot be overstated.
+
+==== Example Use Case
+
+Say for instance, that we have a production cluster onto which we need to onboard applications in a standardized way. We might develop a template for the _projects_ that we will create, including standard `ServiceAccounts` to use for automation and `RoleBindings` to grant the proper privileges to user groups. Additionally, we might want to _deploy_ some common infrastructure (such as Jenkins) to each project, for which we would create another template. For each project that will be instantiated, we will also create a _parameters file_ that can be fed to the template to customize it for each project.
+
+A directory structure for the _infrastructure as code_ repo for this cluster might look like:
+
+[source,bash]
+----
+/repository_root/
+  REAMDE.md # Don't skip Documentation!
+  ... other files & folders ...
+  ./.openshift/
```

Comment:
Is `.openshift` the directory structure that was agreed upon?

Reply:
for now. if it changes that can be a followup